core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -74,16 +73,20 @@
         this.unassignedInfo = unassignedInfo;
         this.allocationId = allocationId;
         this.expectedShardSize = expectedShardSize;
+        this.recoverySource = recoverySource;
         assert expectedShardSize == UNAVAILABLE_EXPECTED_SHARD_SIZE || state == ShardRoutingState.INITIALIZING || state == ShardRoutingState.RELOCATING : expectedShardSize + " state: " + state;
         assert expectedShardSize >= 0 || state != ShardRoutingState.INITIALIZING || state != ShardRoutingState.RELOCATING : expectedShardSize + " state: " + state;
         assert !(state == ShardRoutingState.UNASSIGNED && unassignedInfo == null) : "unassigned shard must be created with meta";
+        assert (state == ShardRoutingState.UNASSIGNED || state == ShardRoutingState.INITIALIZING) == (recoverySource != null) : "recovery source only available on unassigned or initializing shard ";
```

Comment:
it feels a bit weird to me that an unassigned shard has to know it's recovery source. I think this is only needed when we mark as initializing? In the same line - is there any reason not to keep the recovery source when a shard is started? does it hurt to know where it was recovered from?